src/metabase/lib/metadata/jvm.clj

```diff
@@ -409,5 +411,9 @@
   and [[metabase.lib.metadata.protocols/BulkMetadataProvider]]. All operations are cached; so you can use the bulk
   operations to pre-warm the cache if you need to."
   [database-id :- ::lib.schema.id/database]
-  (lib.metadata.cached-provider/cached-metadata-provider
-   (->UncachedApplicationDatabaseMetadataProvider database-id)))
+  (lib.metadata.composed-provider/composed-metadata-provider
+   ;; inovcation tracker has to be the 1st provider in the chain so that all
+   ;; inovcations go through it first
+   (lib.metadata.invocation-tracker/invocation-tracker-provider)
```

Comment:
I think this is risky, and complects `composed-metadata-provider`, invocation tracking, and the QP's metadata handling more than necessary.

I think we should employ the Strategy pattern to full effect here. It's just a protocol, so let's create a new wrapper instead of a mock provider to go in the `composed-metadata-provider`.

```clojure
(deftype InvocationTracker [delegate invocations]
  lib.metadata.protocols/MetadataProvider
  (card [_this card-id]
    (swap! invocations update-in [:metadata/card card-id] (fnil inc 0))
    (lib.metadata/protocols/card deletgate card-id))
  ; ...
  )
```

We can probably write (or find on Clojars) some macro magic to automate that sort of decoration, but it's not hard to just implement it directly when we only have a few.

Reply:
I did that at first but changed to the composed provider because it requires the top-level provider to implement all of the interface that its child implements.

In this case, InvocationTracker will need to implement CachedMetdataProvider and BulkMetadadataProvider.

But maybe I overlooked something and if you think it's a trade-off worth making then that's ok.

I did that in : 268caf6a4d7d31b09cc1499e6bb146d8039d425b